src/modules/landing_target_estimator/LandingTargetEstimator.cpp

```diff
@@ -50,260 +51,1431 @@
+		const sensor_gps_s vehicle_gps_position, targetObsPos *obs)
 {
-	_vehicleLocalPosition_valid = _vehicleLocalPositionSub.update(&_vehicleLocalPosition);
-	_vehicleAttitude_valid = _attitudeSub.update(&_vehicleAttitude);
-	_vehicle_acceleration_valid = _vehicle_acceleration_sub.update(&_vehicle_acceleration);
 
+	// TODO: convert .s_variance_m_s from accuracy to variance
 
-	if (_irlockReportSub.update(&_irlockReport)) { //
-		_new_irlockReport = true;
+	if (_target_mode == TargetMode::Stationary) {
 
-		if (!_vehicleAttitude_valid || !_vehicleLocalPosition_valid || !_vehicleLocalPosition.dist_bottom_valid) {
-			// don't have the data needed for an update
-			return;
+		obs->meas_xyz(0) = -vehicle_gps_position.vel_n_m_s;
+		obs->meas_xyz(1) = -vehicle_gps_position.vel_e_m_s;
+		obs->meas_xyz(2) = -vehicle_gps_position.vel_d_m_s;
+
+		obs->meas_unc_xyz(0) = vehicle_gps_position.s_variance_m_s;
+		obs->meas_unc_xyz(1) = vehicle_gps_position.s_variance_m_s;
+		obs->meas_unc_xyz(2) = vehicle_gps_position.s_variance_m_s;
+
+	} else {
+
+		// If the target is moving, the relative velocity is expressed as the drone verlocity - the target velocity
+		obs->meas_xyz(0) = vehicle_gps_position.vel_n_m_s - target_GNSS_report.vx_rel;
+		obs->meas_xyz(1) = vehicle_gps_position.vel_e_m_s - target_GNSS_report.vy_rel;
+		obs->meas_xyz(2) = vehicle_gps_position.vel_d_m_s - target_GNSS_report.vz_rel;
+
+		float unc = vehicle_gps_position.s_variance_m_s + _gps_target_unc;
+		obs->meas_unc_xyz(0) = unc;
+		obs->meas_unc_xyz(1) = unc;
+		obs->meas_unc_xyz(2) = unc;
+
+		// TODO: uncomment once the mavlink message is updated with covariances
+		// obs->meas_unc_xyz(0) = vehicle_gps_position.s_variance_m_s + target_GNSS_report.cov_vx_rel;
+		// obs->meas_unc_xyz(1) = vehicle_gps_position.s_variance_m_s + target_GNSS_report.cov_vy_rel;
+		// obs->meas_unc_xyz(2) = vehicle_gps_position.s_variance_m_s + target_GNSS_report.cov_vx_rel;
+	}
+
+	if (_target_model == TargetModel::FullPoseCoupled) {
+		// State: [rx, ry, rz, rx_dot, ry_dot, rz_dot, ... ]
+		obs->meas_h_xyz(0, 3) = 1; // x direction
+		obs->meas_h_xyz(1, 4) = 1; // y direction
+		obs->meas_h_xyz(2, 5) = 1; // z direction
+
+	} else {
+		// State: [r, r_dot, ...] --> x, y, z directions are the same (decoupled)
+		obs->meas_h_xyz(0, 1) = 1;
+		obs->meas_h_xyz(1, 1) = 1;
+		obs->meas_h_xyz(2, 1) = 1;
+	}
+
+	obs->timestamp = vehicle_gps_position.timestamp;
+
+	obs->updated_xyz.setAll(true);
+
+	// Keep track of the initial relative velocity
+	_vel_rel_init.timestamp = vehicle_gps_position.timestamp;
+	_vel_rel_init.valid = vehicle_gps_position.vel_ned_valid;
+	_vel_rel_init.vel(0) = -vehicle_gps_position.vel_n_m_s;
+	_vel_rel_init.vel(1) = -vehicle_gps_position.vel_e_m_s;
+	_vel_rel_init.vel(2) = -vehicle_gps_position.vel_d_m_s;
+
+	return true;
+
+}
+
+bool LandingTargetEstimator::processObsTargetGNSS(const landing_target_pose_s target_GNSS_report,
```

Comment:
```suggestion
bool LandingTargetEstimator::processObsTargetGNSS(const landing_target_pose_s &target_GNSS_report,
```
pass structs by const ref

Reply:
Done in https://github.com/JonasPerolini/PX4-Autopilot/commit/ea0c0113b1a7fbe7f4465e657dd69ad93a59c675